Read all data from the device, disable double-encoding, implement more APIs, refactor querying, update README

rytilahti opened the pull request:
The commit messages are pretty self-explanatory, so here's just some extra info:
* My HS110 doesn't seem to send all the data for requests, causing incomplete JSON as response.
```
DEBUG:pyHS100.pyHS100:> (56) {"emeter": {"get_daystat": {"month": 11, "year": 2016}}}
DEBUG:pyHS100.pyHS100:Got chunk 1448
DEBUG:pyHS100.pyHS100:Got chunk 161
DEBUG:pyHS100.pyHS100:Got chunk 0
DEBUG:pyHS100.pyHS100:< (1605) {"emeter":{"get_daystat":{"day_list":[{"year":2016,"month":11,"day":1,"energy":0.404000},
```
* Casting month & year to strings the following, could this be device specific? The older version used by Home Assistant seems to work fine on this case.
```
DEBUG:pyHS100.pyHS100:> (60) {"emeter": {"get_daystat": {"month": "12", "year": "2016"}}}
DEBUG:pyHS100.pyHS100:< (71) {"emeter":{"get_daystat":{"err_code":-3,"err_msg":"invalid argument"}}}
```

rytilahti committed:
Read from socket until no data available, disable double string encoding

HS110 sends sometimes datagrams in chunks especially for get_daystat,
this patch makes it to read until there is no more data to be read.

As json.dumps() does JSON encoding already, there's no need to str()
the year or month either.

rytilahti committed:
Add cli.py, a simple script to query devices for debugging purposes.

kirichkov commented:
Can you give us a dump of the system info for your plug? You could be having different firmware version than mine which leads to those differences.

rytilahti commented:
Threre you go, I masked some information with XXXXXes.
```
DEBUG:pyHS100.pyHS100:< (580) {"system":
{"get_sysinfo":{"err_code":0,
"sw_ver":"1.0.8 Build 151101 Rel.24452",
"hw_ver":"1.0",
"type":"smartplug",
"model":"HS110(EU)"
,"mac":"50:C7:BF:XXXXXX",
"deviceId":"8006588E50AD38930XXXXXXXXXXXXXX",
"hwId":"45E29DA8382494D2EXXXXXXXX",
"fwId":"3A1C9C60B93A090DF074A74960BA60C0",
"oemId":"3D341ECE302C0642C99E31CEXXXXXXXXXXX",
"alias":"Media",
"dev_name":"Wi-Fi Smart Plug With Energy Monitoring",
"icon_hash":"",
"relay_state":1,
"on_time":9921,
"active_mode":"schedule",
"feature":"TIM:ENE",
"updating":0,
"rssi":-51,
"led_off":0,"latitude":XXXXX,"longitude":XXXXXX}}}
```
edit: made it readable
edit: to add, the responses are not always chunked, it just happens now and then (although more often than not it seems).

kirichkov commented:
I did a quick test (using the cli.py file) and your code seems to be working on both python 2 and 3. I think it'd be also good to bump the version to 0.2.2 in /setup.py and merge this.

EDIT: I'm not sure the cli.py should be part of the module, it does help, but it might be better used as an example in the README, rather than a standalone python script.

Thanks for sharing the model info. My plug appears to be the exact same hardware and software version like yours. 
I haven't noticed the issue that you are describing with the chunked data, but I always suspected that socket read/writes are somehow related to the issue me and mweinelt are discussing in #3 .

rytilahti commented:
I agree about cli.py, although maybe it could be placed into a separate dir like examples, and include a shorter usage example in README.md?

Unfortunately I don't have any historical data left as I moved my plug and cleared the device, but if you create a loop for daily you'd probably see the device delivering incomplete packets sometimes.

Unfortunately I have no insight what may cause the behavior in #3, the sockets seem to be closed properly, but in case it's not also done on the plug's end, maybe it runs out of file descriptors? One could test that by doing lots of queries and see if it stops being responsive.

rytilahti committed:
allow easier importing with from pyHS100 import SmartPlug

rytilahti committed:
move cli.py to examples, add short usage into README.md

rytilahti committed:
Implement more available APIs, refactor querying code.

This commit adds access to new properties, both read & write,  while keeping the old one (mostly) intact.
Querying is refactored to be done inside _query_helper() method,
which unwraps results automatically and rises SmartPlugException() in case of errors.
Errors are to be handled by clients.

New features:
* Setting device alias (plug.alias = "name")
* led read & write
* icon read (doesn't seem to return anything without cloud support at least), write API is not known, throws an exception currently
* time read (returns datetime), time write implemented, but not working even when no error is returned from the device
* timezone read
* mac read & write, writing is untested for now.

Properties for easier access:
* hw_info: return hw-specific elements from sysinfo
* on_since: pretty-printed from sysinfo
* location: latitude and longitued from sysinfo
* rssi: rssi from sysinfo

rytilahti committed:
Update README.md with examples of available features.

rytilahti commented:
Uhuh, I may have made a mess by working on the same initial branch for new set of patches. If that's okay then fine, if not, I'll have to figure out how to rebase and split these changes. :)

edit: looks like I messed up some changes from others... Will have to check that out when I get some time for that. For now this can be put in pending mode.

mweinelt commented:
Can you please rebase your commits onto the current state of the master branch?

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -318,12 +318,20 @@ def query(host, request, port=9999):
+            chunk = sock.recv(4096)
+            buffer += chunk
+            #_LOGGER.debug("Got chunk %s" % len(chunk))
+            if len(chunk) == 0:
```
```
if not chuck:
```

rytilahti replied:
Good catch, didn't recall for sure whether empty b'' would've matched that, but it does.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -318,12 +318,20 @@ def query(host, request, port=9999):
 
         sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         sock.connect((host, port))
+        _LOGGER.debug("> (%i) %s" % (len(request), request))
```
String interpolation is deprecated, use `str.format()` instead.

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -318,12 +318,20 @@ def query(host, request, port=9999):
 
-        response = TPLinkSmartHomeProtocol.decrypt(buffer)
+        response = TPLinkSmartHomeProtocol.decrypt(buffer[4:])
+        _LOGGER.debug("< (%i) %s" % (len(response), response))
```
Interpolation here as well.

rytilahti replied:
Done.

mweinelt commented on cli.py:
```diff
@@ -0,0 +1,26 @@
+logging.info("Sysinfo: %s" % hs.get_sysinfo())
+has_emeter = hs.has_emeter
+if has_emeter:
+  logging.info("== Emeter ==")
```
Please indent with 4 spaces.

rytilahti replied:
Done.

mweinelt commented on cli.py:
```diff
@@ -0,0 +1,26 @@
+logging.basicConfig(level=logging.DEBUG)
+
+if len(sys.argv) < 2:
+  print("%s <ip>" % sys.argv[0])
```
Please indent with 4 spaces.

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -60,7 +65,30 @@ def __init__(self, ip_address):
+
+        result = response[target][cmd]
+        if result["err_code"] != 0:
+            raise SmartPlugException("Error on %s.%s: %s" % (target, cmd, result))
```
String interpolation

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Sets the device name aka alias.
+        :param alias: New alias (name)
+        """
+
```
remove empty line here

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Returns the state of the led.
+        :return:
+        """
+        return {"led": 1 - self.sys_info["led_off"]}
```
Why a `dict` here and not simply a boolean?
`return bool(self.sys_info['led_off'])`

rytilahti replied:
Good catch, no idea, changed.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Sets the state of the led (night mode)
+        :param state: 1 to set led on, 0 to set led off
+        """
+
```
remove empty line

rytilahti replied:
Done. Would it be good idea also to return empty :returns ? My argument being, when the description already tells what it does, e.g. in case of led().

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        :param icon:
+        """
+
+        raise SmartPlugNotImplementedException()
```
I'm not sure there is much of a benefit to roll our own Exception in this case. I'd say lets simply raise `NotImplementedError`.

rytilahti replied:
I was thinking that it allows clients to catch SmartPlugException (as that inherits from it) without needing separate handling. Does that make sense to you?
edit: changed that. it doesn't really bring that much and causes just mess.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        """
+
+        response = self._query_helper("time", "get_time")
+        ts = datetime.datetime(response["year"], response["month"], response["mday"], response["hour"], response["min"], response["sec"])
```
just return here, there is no need to bind this to a name.

rytilahti replied:
I usually prefer to set it up like that because it allows easier introspection with a debugger when needed, but done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Returns timezone information
+        :return:
+        """
+
```
remove empty line here

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        :param ts: New timestamp
+        :return:
+        """
+
```
remove empty line here

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        :return:
+        """
+        keys = ["sw_ver", "hw_ver", "mac", "hwId", "fwId", "oemId", "dev_name"]
+        return {key:self.sys_info[key] for key in keys}
```
add a whitespace after the colon.

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Location of the device, as read from sysinfo
+        :return:
+        """
+
```
remove the empty line

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Returns WiFi signal strenth (rssi)
+        :return: rssi
+        """
+
```
remove the empty line

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Sets new mac address
+        :param mac: mac in hexadecimal with colons, e.g. 01:23:45:67:89:ab
+        """
+
```
remove the empty line

rytilahti replied:
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -285,6 +268,156 @@ def identify(self):
+        Content for hash and icon are unknown.
+        :param icon:
+        """
+
```
remove the empty line

rytilahti replied:
Done.

rytilahti committed:
Handle comments from mweinelt

rytilahti commented:
Thanks for the review and insightful comments :) I hope I managed to clean-up the mess I made by merging through github in my repo..

mweinelt commented:
I wouldn't expect anyone to implement a call to that function when it's
not implemented in the first place.

On 12/02/2016 08:59 PM, Teemu R. wrote:
> *@rytilahti* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In pyHS100/pyHS100.py <https://github.com/GadgetReactor/pyHS100/pull/11>:
> 
>> +    def icon(self):
> +        """
> +        Returns device icon
> +        Note: this doesn't seem to work when not using the cloud service, not tested with it either.
> +        :return:
> +        """
> +        return self._query_helper("system", "get_dev_icon")
> +
> +    @icon.setter
> +    def icon(self, icon):
> +        """
> +        Content for hash and icon are unknown.
> +        :param icon:
> +        """
> +
> +        raise SmartPlugNotImplementedException()
> 
> I was thinking that it allows clients to catch SmartPlugException (as
> that inherits from it) without needing separate handling. Does that make
> sense to you?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/GadgetReactor/pyHS100/pull/11>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAICD09Nf_DXFOm-FRXKP1qfN2RheUX7ks5rEHiZgaJpZM4LB0ly>.
>

rytilahti committed:
Refactor state handling, use booleans instead of strings

rytilahti commented:
Agreed. The newest commit simplifies state handling by converting from strings to bools, e.g.
plug.state = "ON" is now plug.state = True

A couple of questions:
* As all "active" actions do throw an exception in case of failure, it's probably a good idea to add :raises SmartPlugException to those for clients?
* In case of description containing already what it returns, is it necessary to have that type of boilerplate there in :returns too?

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -23,8 +23,6 @@
 
 # switch states
-SWITCH_STATE_ON = 'on'
-SWITCH_STATE_OFF = 'off'
```
Let's keep those and let people refer to them, just in case we decide to change the representation again.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -93,18 +91,16 @@ def state(self):
+            return bool(relay_state)
         else:
-            _LOGGER.warning("Unknown state %s returned.", relay_state)
+            _LOGGER.warning("Unknown state %s returned.".formaat(relay_state))
```
> formaat

kirichkov commented:
The reason why "ON" and "OFF" are used instead of Booleans is because the module is modeled after the D-Link module handling D-Link's smart plugs.

Also from readability stand point if we are to replace ON/OFF with True/False the name of state should be changed to something like `is_on`. I'm fine with keeping the internal representation like a Boolean, but the state property should still return "ON" / "OFF" first for backwards compatibility and second for readability. Two new methods can be added `is_on` and `is_off` Which return True/False.

mweinelt commented:
Yes, we should document `raises:` where necessary, so developers quickly see what they have to expect.

Regarding `returns:`, I think it should be added everywhere, where return is not `None`. The reasoning is consistency.

rytilahti committed:
Fix issues raised during the review.

Following issues are addressed by this commit:
* All API is more or less commented (including return types, exceptions, ..)
* Converted state to use
* Added properties is_on, is_off for those who don't want to check against strings.
* Handled most issues reported by pylint.
* Adjusted _query_helper() to strip off err_code from the result object.
* Fixed broken format() syntax for string formattings.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -54,12 +59,40 @@ def __init__(self, ip_address):
+
+        :param target: Target system {system, time, emeter, ..}
+        :param cmd: Command to execute
+        :param arg: JSON object passed as parameter to the command, defualts to {}
```
Also typo "defualts"

rytilahti replied:
Fixed "defualts".

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -316,12 +492,20 @@ def query(host, request, port=9999):
+        while True:
+            chunk = sock.recv(4096)
+            buffer += chunk
+            #_LOGGER.debug("Got chunk %s".format(len(chunk)))
```
Really? Can fix though, I don't see it helping much in case of such debug messages, which may sometimes be useful to have. Better strip it off completely?

mweinelt replied:
Yes, better drop it.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        :rtype: dict
+        """
+
+        return {"latitude": self.sys_info["latitude"], "longitude": self.sys_info["longitude"]}
```
Fixed.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        :return: datetime for on since
+        :rtype: datetime
+        """
+        return datetime.datetime.now() - datetime.timedelta(seconds=self.sys_info["on_time"])
```
Found that more readable than with a newline to be honest, but fixed.

mweinelt replied:
We should adhere to PEP8 to guarantee consistently formatted code, and the line limit ensures even lower resolution devices can display the code in a reasonable manner.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        :raises SmartPlugException: on error
+        """
+        raise NotImplementedError("Setting time does not seem to work on HS110 although it returns no error.")
+        """ here just for the sake of completeness / if someone figures out why it doesn't work.
```
Um, pep8's 79 chars is more or less a recommendation, but of controversial sort. Can fix this and one below though if wanted.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        :raises NotImplemented: when not implemented.
+        :raises SmartPlugException: on error
+        """
+        raise NotImplementedError("Setting time does not seem to work on HS110 although it returns no error.")
```
See above.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        :raises SmartPlugException: on error
+        """
+        response = self._query_helper("time", "get_time")
+        return datetime.datetime(response["year"], response["month"], response["mday"], response["hour"], response["min"], response["sec"])
```
Fixed.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        """
+        raise NotImplementedError("Values for this call are unknown at this point.")
+        # here just for the sake of completeness
+        # self._query_helper("system", "set_dev_icon", {"icon": "", "hash": ""})
```
See above mention about pep8.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+        :raises NotImplementedError: when not implemented
+        :raises SmartPlugError: on error
+        """
+        raise NotImplementedError("Values for this call are unknown at this point.")
```
See above.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +297,168 @@ def identify(self):
+    def icon(self):
+        """
+        Returns device icon
+        Note: this doesn't seem to work when not using the cloud service, not tested with it either.
```
Fixed.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -217,18 +236,14 @@ def get_emeter_monthly(self, year=datetime.datetime.now().year):
-
-        if response['err_code'] != 0:
-            return False
+        response = self._query_helper("emeter", "get_monthstat", {'year': year})
```
Done.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -199,14 +225,7 @@ def get_emeter_daily(self, year=None, month=None):
-
-        if response['err_code'] != 0:
-            return False
+        response = self._query_helper("emeter", "get_daystat", {'month': month, 'year': year})
```
Done.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -54,12 +59,40 @@ def __init__(self, ip_address):
+
+        result = response[target][cmd]
+        if result["err_code"] != 0:
+            raise SmartPlugException("Error on {}.{}: {}".format(target, cmd, result))
```
I'd prefer to keep this again for the sake of readability. Moving the call to format to a new line or splitting up the parameter list doesn't really improve reaadability.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -22,6 +22,11 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+class SmartPlugException(Exception):
```
Done.

mweinelt commented on pyHS100/pyHS100.py:
```diff
@@ -54,12 +59,40 @@ def __init__(self, ip_address):
+    def _query_helper(self, target, cmd, arg={}):
+        """
+        Query helper, returns unwrapped result object.
+        Raises SmartPlugException in case of error reported by the plug.
```
Use `:raises` and move after last param for consistency.

rytilahti replied:
It's later on there. This is just the descriptive string for it.

rytilahti commented on pyHS100/__init__.py:
```diff
@@ -1 +1 @@
-
+from pyHS100.pyHS100 import SmartPlug
```
False alarm. import pyHS100 allows using pyHS100.SmartPlug as well as from pyHS100 import SmartPlug for library users. Can be reverted though, if from pyHS100.pyHS100 import SmartPlug is more wanted.

rytilahti commented:
Fixed quite a few things, as earlier, commit message contains the changes. Btw, pylint complains that pyHS100 is not a valid name :-)

rytilahti committed:
Fix ci woes plus one typo.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +301,173 @@ def identify(self):
+        :rtype: datetime
+        """
+        return datetime.datetime.now() - \
+               datetime.timedelta(seconds=self.sys_info["on_time"])
```
Again for readability's sake. If wanted, we could import datetime & timedelta directly and avoid typing datetime twice/at all, if wanted.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +301,173 @@ def identify(self):
+        :raises SmartPlugException: on error
+        """
+        response = self._query_helper("time", "get_time")
+        return datetime.datetime(response["year"], response["month"], response["mday"],
```
Please, it's already splitted. Changing response to rsp or r doesn't improve readability.

rytilahti commented on pyHS100/pyHS100.py:
```diff
@@ -283,6 +301,173 @@ def identify(self):
+        """
+        Returns device icon
+
+        Note: this doesn't seem to work (at least) when not using the cloud service.
```
Better remove this altogether, if this is an issue. It's for there for devels to see/note that there may be an issue there..

mweinelt commented:
Yes, the name should be PyHS100 instead - or at least start with a captical character. But let's keep that for another PR, this one's large enough already.

rytilahti commented:
Yeah, it's my failure to mess up that much with using proper feature branches, wasn't thinking it would gain so many commits in the end.
edit: btw, it'd make sense to have unit tests for the state machine & error handling functions, although I wouldn't like to have that either in this commit series, but do it separately.

mweinelt commented:
Agreed, unit testing makes total sense.

rytilahti commented:
On related note, perhaps LED api should also follow the "legacy" ON/OFF for state values, even if I don't really like strings as return values/parameters?

mweinelt commented:
I agree True/False should be used, not strings.

rytilahti committed:
Do initialization after changing device properties, fix nits.

rytilahti commented:
Adjusted led setter to use True/False instead of ints, fix some styling issues (keep whitespaces in docstrings between headline & other elements. Outed initialization out from __init__() and force-calling it after all API calls changing the state of the device.

I don't really like the initialize solution, maybe sysinfo content shouldbe kept locally and just updated based on responses from the plug. Opinions?

mweinelt commented:
@GadgetReactor Can we get this merged?

GadgetReactor commented:
thanks!

rytilahti commented:
Thanks for merging, next step is to clean the code base a bit & prepare for easier testing :-)
